File: problems/tier1-foundation/004-vending-machine/solution.js
// Vending Machine — State pattern reference solution (JavaScript).

const VMState = {
  IDLE: "Idle",
  PAYMENT_PENDING: "PaymentPending",
  DISPENSING: "Dispensing",
  MAINTENANCE: "Maintenance",
};

class VendingMachine {
  constructor() {
    this.state = VMState.IDLE;
    this.selectedItem = "";
    this.insertedMoney = 0.0;
    this.inventory = new Map();
    this.operatorPin = "1234";
    this._stockDefaults();
  }

  _stockDefaults() {
    this.inventory.set("Cola", { name: "Cola", price: 25.0, quantity: 5 });
    this.inventory.set("Chips", { name: "Chips", price: 15.0, quantity: 3 });
  }

  reset() {
    this.state = VMState.IDLE;
    this.selectedItem = "";
    this.insertedMoney = 0.0;
    this._stockDefaults();
  }

  selectItem(item) {
    if (this.state !== VMState.IDLE) return;
    if (this.inventory.has(item) && this.inventory.get(item).quantity > 0) {
      this.selectedItem = item;
      this.state = VMState.PAYMENT_PENDING;
    }
  }

  insertMoney(amount) {
    if (this.state !== VMState.PAYMENT_PENDING) return;
    this.insertedMoney += amount;
    if (this.insertedMoney >= this.inventory.get(this.selectedItem).price) {
      this.state = VMState.DISPENSING;
    }
  }

  dispense() {
    if (this.state !== VMState.DISPENSING) return;
    this.inventory.get(this.selectedItem).quantity -= 1;
    this.insertedMoney = 0.0;
    this.selectedItem = "";
    this.state = VMState.IDLE;
  }

  cancel() {
    if (this.state !== VMState.PAYMENT_PENDING) return;
    this.insertedMoney = 0.0;
    this.selectedItem = "";
    this.state = VMState.IDLE;
  }

  enterMaintenance(pin) {
    if (pin === this.operatorPin && this.state === VMState.IDLE) {
      this.state = VMState.MAINTENANCE;
    }
  }

  exitMaintenance(pin) {
    if (pin === this.operatorPin && this.state === VMState.MAINTENANCE) {
      this.state = VMState.IDLE;
    }
  }

  restock(item, qty) {
    if (this.state !== VMState.MAINTENANCE) return;
    if (!this.inventory.has(item)) {
      this.inventory.set(item, { name: item, price: 0.0, quantity: 0 });
    }
    this.inventory.get(item).quantity += qty;
  }
}

let _g_vm = new VendingMachine();

function reset_service() {
  _g_vm = new VendingMachine();
}

function reset() {
  _g_vm.reset();
}

function getState() {
  return _g_vm.state;
}

function selectItem(item) {
  _g_vm.selectItem(item);
}

function insertMoney(amount) {
  _g_vm.insertMoney(amount);
}

function dispense() {
  _g_vm.dispense();
}

function cancel() {
  _g_vm.cancel();
}

function enterMaintenance(pin) {
  _g_vm.enterMaintenance(pin);
}

function exitMaintenance(pin) {
  _g_vm.exitMaintenance(pin);
}

function restock(item, qty) {
  _g_vm.restock(item, qty);
}

function vm_get_quantity(item) {
  if (!_g_vm.inventory.has(item)) return -1;
  return _g_vm.inventory.get(item).quantity;
}

function vm_get_inserted_money() {
  return _g_vm.insertedMoney;
}

function vm_get_selected_item() {
  return _g_vm.selectedItem;
}

module.exports = {
  VendingMachine,
  reset_service,
  reset,
  getState,
  selectItem,
  insertMoney,
  dispense,
  cancel,
  enterMaintenance,
  exitMaintenance,
  restock,
  vm_get_quantity,
  vm_get_inserted_money,
  vm_get_selected_item,
};

File: problems/tier1-foundation/004-vending-machine/solution.py
"""Vending Machine — State pattern reference solution (Python)."""


class VMState:
    IDLE = "Idle"
    PAYMENT_PENDING = "PaymentPending"
    DISPENSING = "Dispensing"
    MAINTENANCE = "Maintenance"


class VendingMachine:
    def __init__(self):
        self.state = VMState.IDLE
        self.selectedItem = ""
        self.insertedMoney = 0.0
        self.inventory = {}
        self.operatorPin = "1234"
        self._stockDefaults()

    def _stockDefaults(self):
        self.inventory["Cola"] = {"name": "Cola", "price": 25.0, "quantity": 5}
        self.inventory["Chips"] = {"name": "Chips", "price": 15.0, "quantity": 3}

    def reset(self):
        self.state = VMState.IDLE
        self.selectedItem = ""
        self.insertedMoney = 0.0
        self._stockDefaults()

    def selectItem(self, item):
        if self.state != VMState.IDLE:
            return
        if item in self.inventory and self.inventory[item]["quantity"] > 0:
            self.selectedItem = item
            self.state = VMState.PAYMENT_PENDING

    def insertMoney(self, amount):
        if self.state != VMState.PAYMENT_PENDING:
            return
        self.insertedMoney += amount
        if self.insertedMoney >= self.inventory[self.selectedItem]["price"]:
            self.state = VMState.DISPENSING

    def dispense(self):
        if self.state != VMState.DISPENSING:
            return
        self.inventory[self.selectedItem]["quantity"] -= 1
        self.insertedMoney = 0.0
        self.selectedItem = ""
        self.state = VMState.IDLE

    def cancel(self):
        if self.state != VMState.PAYMENT_PENDING:
            return
        self.insertedMoney = 0.0
        self.selectedItem = ""
        self.state = VMState.IDLE

    def enterMaintenance(self, pin):
        if pin == self.operatorPin and self.state == VMState.IDLE:
            self.state = VMState.MAINTENANCE

    def exitMaintenance(self, pin):
        if pin == self.operatorPin and self.state == VMState.MAINTENANCE:
            self.state = VMState.IDLE

    def restock(self, item, qty):
        if self.state != VMState.MAINTENANCE:
            return
        if item not in self.inventory:
            self.inventory[item] = {"name": item, "price": 0.0, "quantity": 0}
        self.inventory[item]["quantity"] += qty


_g_vm = VendingMachine()


def reset_service():
    global _g_vm
    _g_vm = VendingMachine()


def reset():
    _g_vm.reset()


def getState():
    return _g_vm.state


def selectItem(item):
    _g_vm.selectItem(item)


def insertMoney(amount):
    _g_vm.insertMoney(amount)


def dispense():
    _g_vm.dispense()


def cancel():
    _g_vm.cancel()


def enterMaintenance(pin):
    _g_vm.enterMaintenance(pin)


def exitMaintenance(pin):
    _g_vm.exitMaintenance(pin)


def restock(item, qty):
    _g_vm.restock(item, qty)


def vm_get_quantity(item):
    if item not in _g_vm.inventory:
        return -1
    return _g_vm.inventory[item]["quantity"]


def vm_get_inserted_money():
    return _g_vm.insertedMoney


def vm_get_selected_item():
    return _g_vm.selectedItem
